Use STEitherAmount as a variant of STAmount and STMPTAmount

gregtatcam opened the pull request:
Introduce STEitherAmount to represent STAmount and STMPTAmount as a variant. Accessing specific amount field requires a call to get<T>(). 
Add amendment guards to all transactions with STI_AMOUNT field.
Extend `SOElement` with enum `SOESupportMPT` field. `TxFormats::TxFormats()` is used to declare amount fields supporting both `STAmount` and `STMPTAmount`. This provides an additional run-time type-safety.

gregtatcam committed:
Use STEitherAmount as a variant of STAmount and MPTAmount

gregtatcam committed:
Split AMOUNT fields into two types SF_AMOUNT (don't support MPT)
and SF_EITHER_AMOUNT (support MPT).
Add TypedFieldAmount to declare Amount fields.
Extend STObject with getters/setters for SF_AMOUNT fields.

gregtatcam committed:
Remove STBase and CountedObject inheritance from STAmount and STMPTAmount

gregtatcam committed:
Add invalid MPT amount check in STTx::passesLocalChecks()
to provide one single check for all tx.
Update unit-tests.
Fix payment for unauthorized holder

gregtatcam committed:
Consolidate errors. Remove redundant checks.

shawnxie999 commented on include/xrpl/protocol/STEitherAmount.h:
```diff
@@ -0,0 +1,313 @@
+get(auto&& amount)
+{
+    using TAmnt = std::decay_t<decltype(amount)>;
+    if constexpr (std::is_same_v<TAmnt, STEitherAmount>)
```
is this to support `get<STEitherAmount>`? I don't think it is being used anywhere

gregtatcam replied:
It's used to get either `STAmount` or `STMPTAmount`.  See the concept `ValidAmountType`.

shawnxie999 commented on include/xrpl/protocol/STEitherAmount.h:
```diff
@@ -0,0 +1,313 @@
+    if constexpr (std::is_same_v<TAmnt, STEitherAmount>)
+    {
+        if constexpr (std::is_lvalue_reference_v<decltype(amount)>)
+            return amount.template get<T>();
```
would this call the either one of two `get()`?:

```
template <ValidAmountType T>
T const&
STEitherAmount::get() const
{
    if (std::holds_alternative<T>(amount_))
        return std::get<T>(amount_);
    Throw<std::logic_error>("Invalid STEitherAmount conversion");
}

template <ValidAmountType T>
T&
STEitherAmount::get()
{
    if (std::holds_alternative<T>(amount_))
        return std::get<T>(amount_);
    Throw<std::logic_error>("Invalid STEitherAmount conversion");
```
It seems like both of them would throw.

gregtatcam replied:
One is const and the other one is not.

gregtatcam replied:
And, yes one of them is going to be called and will throw if the requested type doesn't match the type that the variant holds.

shawnxie999 commented on include/xrpl/protocol/STEitherAmount.h:
```diff
@@ -0,0 +1,313 @@
+        },
+        lhs.getValue(),
+        rhs.getValue());
+    if (lhs.isIssue() == rhs.isIssue())
```
is this code reachable?

gregtatcam replied:
Thanks. It's not reachable. I was debating what version to use and forgot to remove it.

gregtatcam committed:
Fix serialization. Extend unit-tests.

gregtatcam committed:
Refactor Payment and Clawback transactors
to have separate functions for MPT/IOU
handling since there is little overlap
between them.
Remove the check for DeletableAccounts

shawnxie999 commented on src/test/app/MPToken_test.cpp:
```diff
@@ -951,9 +1372,9 @@ class MPToken_test : public beast::unit_test::suite
             env.close();
 
-            // TODO: uncomment after stamount changes
-            // env(claw(alice, mpt(maxMPTokenAmount), bob), ter(temBAD_AMOUNT));
```
dont we still want to check for maxMPTokenAmount + 1? what error code does it return in this case?

shawnxie999 commented on src/libxrpl/basics/MPTAmount.cpp:
```diff
@@ -0,0 +1,85 @@
+MPTAmount&
+MPTAmount::operator+=(MPTAmount const& other)
+{
+    value_ += other.value();
```
does it ever have to check overflow here?

shawnxie999 replied:
nevermind, its checked in the constructor

shawnxie999 commented on src/test/protocol/STTx_test.cpp:
```diff
@@ -1344,22 +1344,22 @@ class STTx_test : public beast::unit_test::suite
             0x73, 0x00, 0x81, 0x14, 0x00, 0x10, 0x00, 0x00, 0x00, 0x00, 0x00,
             0x00, 0x65, 0x24, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
             0xe5, 0xfe, 0xf3, 0xe7, 0xe5, 0x65, 0x24, 0x00, 0x00, 0x00, 0x00,
-            0x20, 0x1e, 0x00, 0x6f, 0x00, 0x00, 0x20, 0x1f, 0x03, 0xf6, 0x00,
```
what's  changing this for?

gregtatcam replied:
I took this change from Howard's MPT branch. One of the tests fails without it.

shawnxie999 commented on include/xrpl/protocol/SField.h:
```diff
@@ -520,7 +551,7 @@ extern SF_UINT256 const sfHookNamespace;
 
 // currency amount (common)
-extern SF_AMOUNT const sfAmount;
+extern SF_EITHER_AMOUNT const sfAmount;
```
why wouldn't we need to change the other Amount types like sfBalance to SF_EITHER_AMOUNT?

gregtatcam replied:
`SF_EITHER_AMOUNT` is used to indicate that this field can be either `STAmount` or `STMPTAmount`. `SF_AMOUNT` indicates that the field can be only `STAmount`. Only `sfAmount` supports MPT in the first release. What this actually does is this: when you access `sfAmount` as `tx[sfAmount]` then it returns `STEitherAmount` and you have explicitly get the amount with `get<>()`. If you access `sfBalance` as `tx[sfBalance]` then it returns `STAmount`.

shawnxie999 replied:
Yes I understand that. But don't we already have something like

shawnxie999 replied:
Yes I understand that, but don't we already have something like this inside the TxFormat.cpp?
`{sfAmount, soeREQUIRED, soeMPTSupported},`

Or i guess this one is compile-type check where as the SOElement is runtime.

gregtatcam replied:
TxFormat defines a run-time check, which is done in `invalidMPTAmountInTx()`. `SF_AMOUNT` and `SF_EITHER_AMOUNT` declare different types for the fields that don't support and support MPT. These types provide overload for `STObject::operator[]()`, which returns either `STAmount` or `STEitherAmount` depending on the field type. This is compile time. For instance, this is going to fail for `sfAmount`: STAmount const amount = tx[sfAmount]` because `sfAmount` supports MPT.

gregtatcam committed:
Fix Linux build

seelabs commented on include/xrpl/protocol/MPTIssue.h:
```diff
@@ -66,6 +68,25 @@ operator!=(MPTIssue const& lhs, MPTIssue const& rhs)
+getMPT(uint192 const&);
+
+inline MPT
+badMPT()
```
It would be better if we didn't need a sentinel value for a bad mpt. Is this really needed?

seelabs commented on include/xrpl/protocol/SField.h:
```diff
@@ -68,6 +68,7 @@ class STCurrency;
     STYPE(STI_UINT128, 4)                         \
     STYPE(STI_UINT256, 5)                         \
     STYPE(STI_AMOUNT, 6)                          \
+    STYPE(STI_EITHER_AMOUNT, 6)                   \
```
I see this duplicates st amount, and I think I know why it's done. Could we remove STI_AMOUNT?

seelabs commented on include/xrpl/protocol/SField.h:
```diff
@@ -346,7 +376,8 @@ using SF_UINT384 = TypedField<STBitString<384>>;
 using SF_ACCOUNT = TypedField<STAccount>;
-using SF_AMOUNT = TypedField<STAmount>;
+using SF_AMOUNT = TypedFieldAmount<SFieldMPT::No>;
+using SF_EITHER_AMOUNT = TypedFieldAmount<SFieldMPT::Yes>;
```
This seems misleading to me. The template parameter controls whether the type can contain an mpt, but it doesn't necessarily need to hold an mpt - it can still hold an stamount. Maybe rename "Yes" to something else? `MaybeMPT` or somesuch?

seelabs commented on include/xrpl/protocol/STAmount.h:
```diff
@@ -23,21 +23,18 @@
-concept AssetType = std::is_same_v<A, Issue> || std::is_same_v<A, MPT> ||
-    std::is_same_v<A, Asset> || std::is_convertible_v<A, Issue> ||
-    std::is_convertible_v<A, Asset>;
+struct int64_tag_t
```
Consider narrowing the scope of this declaration.

seelabs commented on include/xrpl/protocol/STAmount.h:
```diff
@@ -49,15 +46,15 @@ concept AssetType = std::is_same_v<A, Issue> || std::is_same_v<A, MPT> ||
 // High 8 bits are (offset+142), legal range is, 80 to 22 inclusive
 // Low 56 bits are value, legal range is 10^15 to (10^16 - 1) inclusive
-class STAmount final : public STBase, public CountedObject<STAmount>
+class STAmount final
```
Since this no longer inherits from STBase consider renaming it.

seelabs commented on include/xrpl/protocol/STAmount.h:
```diff
@@ -76,104 +73,59 @@ class STAmount final : public STBase, public CountedObject<STAmount>
-        std::uint64_t mantissa = 0,
-        int exponent = 0,
-        bool negative = false);
+    STAmount(std::int64_t mantissa, int64_tag_t);
```
This needs a comment. I know the existing code is way undercommented, but new code should be better documented. (for this and other changes in this changeset, but this is the first one that caught my eye)

seelabs commented on include/xrpl/protocol/STObject.h:
```diff
@@ -53,12 +54,15 @@ throwFieldNotFound(SField const& field)
 class STObject : public STBase, public CountedObject<STObject>
 {
+    template <typename T, SFieldMPT M>
+    using ValueType = std::conditional<M == SFieldMPT::No, STAmount, T>::type;
```
It would be better if STObject didn't know anything about STAmount.

seelabs commented on include/xrpl/protocol/UintTypes.h:
```diff
@@ -46,12 +46,6 @@ class NodeIDTag
-public:
-    explicit MPTTag() = default;
-};
-
```
I'll leave a comment here, but around line 58 there's the following code:

```
 /** MPT is a 192-bit hash representing MPTID. */
 using MPT = std::pair<std::uint32_t, AccountID>;
 ```
For better type safety, consider making a class for MPT. I'm also not sure this is the best place for this declaration.

Also, around line 69 is the following declaration:
```
 MPT const&
 noMPT();
```
It would be better if we didn't need these sentinal classes. What is the purpose of this? I don't see it used.

seelabs reviewed (approved):
I think the approach is fine. :+1: from me. I did not do a detailed review at all, I'm restricted myself to high-level considerations. However, if I did notice anything while I looked at the code I made a note and left a comment. That doesn't mean the comment is important, they are just things I noticed and wanted to point out.

gregtatcam committed:
Address reviewer's feedback

* Replace MPT:pair<uint32,AccountID> with MPTID:uint192
* Remove SFieldMPT
* Replated TypedFieldAmount with generic
  TypedFieldVariant to handle STEitherAmount
* Update comments

gregtatcam committed:
Merge with latest develop

gregtatcam committed:
Fix clang-format

gregtatcam committed:
Resolve conflicts with mpt-v1

gregtatcam committed:
Fix clang-format
